### DISCUSSION TITLE
Example of a retry mechanism for connecting and subscribing

### DISCUSSION THREAD

**Discussion by h0jeZvgoxFepBQ2C** [nats-io/nats.ws#193] [Q&A]
Hi,

we are currently trying to implement nats.ws which should cover following scenarios:

1) connect in any case (if nats server is offline, retry, if online just connect). also retry if first connection try fails.
2) reconnect indefinately in case connection goes down
3) subscribe to a channel (also after reconnect or if connection was down on first try)

For this usecase we have following questions:

1) We are not sure whats the best syntax for this behaviour with the await notation.
2) We are not sure if we can subscribe for a channel, even when no inital connection could be established?
3) Do we need to resubcribe again after the connection reconnects? 

We also tried the ```waitOnFirstConnect``` but then our connection attempts didnt do anything at all.. As far as we understood ```waitOnFirstConnect``` just retries to connect instead of throwing a timeout error f.e.?

Would be great if you could help us out here!

**Comment by aricart** [comment]
waitOnFirstConnect will continue to retry regardless of the maxReconnects setting. In your case I think you are looking for waitOnFirstConnect and maxReconnects to be -1.

The client won't be able to subscribe or do anything until a connection is established. After a connect, if you disconnect, the client will attempt to reconnect forever, when it does, it will re-establish the subscriptions etc. Note that when the client disconnects, any data that was outbound will be dropped. If you need to know if you published, etc, you'll need to use request reply - on disconnect, your requests will reject.
